lib/Dialect/FIRRTL/Transforms/LowerAnnotations.cpp

```diff
@@ -473,12 +476,36 @@ LogicalResult LowerAnnotationsPass::applyAnnotation(DictionaryAttr anno,
   return success();
 }
 
+/// Convert consumed SourceAnnotation and SinkAnnotation into WiringProblems,
+/// using the pin attribute as newNameHint
+LogicalResult LowerAnnotationsPass::legacyToWiringProblems(ApplyState &state) {
+  for (const auto &problem : state.legacyWiringProblems) {
+    auto pin = cast<StringAttr>(problem.first);
+
+    if (!problem.second.source) {
+      return mlir::emitError(state.circuit.getLoc())
+             << "Unable to resolve source for pin: " << pin;
+    }
+    if (problem.second.sinks.empty()) {
+      return mlir::emitError(state.circuit.getLoc())
+             << "Unable to resolve sink(s) for pin: " << pin;
+    }
+
+    for (const auto &sink : problem.second.sinks) {
+      state.wiringProblems.push_back(
+          {problem.second.source, sink, pin.str(), true});
+    }
+  }
+  return success();
+}
+
 /// Modify the circuit to solve and apply all Wiring Problems in the circuit.  A
-/// Wiring Problem is a mapping from a source to a sink that should be connected
-/// via a RefType.  This uses a two-step approach.  First, all Wiring Problems
-/// are analyzed to compute pending modifications to modules.  Second, modules
-/// are visited from leaves to roots to apply module modifications.  Module
-/// modifications include addings ports and connecting things up.
+/// Wiring Problem is a mapping from a source to a sink that can be connected
+/// via a real Type or RefType as requested.  This uses a two-step approach.
```

Comment:
FWIW non-reftype is "base" not "real", but maybe they should've been called real instead :smile:.  Calling these "real" ports makes sense to me.